### PR TITLE
GEOM: Implement Point and Vector

### LINKED ISSUE #956: Geom: Separate or consolidate `Point` and `Vector`
### Description of the task
Depending on #955, either implement a separate `Point` class or remove it. Either way, remove the `typedef`.

### Acceptance criteria

- [ ] Separate/consolidate `Point` and `Vector`
- [ ] Pass existing tests

### Blocked By

#955

### PR DESCRIPTION
**Pull request by scveloso**
### Description
- Created `new_geom` directory to house components of the geom redesign.
- Implement `Point` and `Vector` separation as a part of the geom redesign. 
- Copied old `Angle` implementation to `new_geom` and updated its usages of `Angle` to `const Angle&` to address #813. Will be doing this periodically when working on the geom redesign.

### Testing Done
- New unit tests for Point and Vector
- Old unit tests for Angle copied to `new_geom` still pass

### Resolved Issues
resolves #956 

### Length Justification
- Easier to do Point and Vector in one go as almost all of the logic is the old logic that was in `geom/point.h`

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/new_geom/angle.h [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+    static constexpr Angle full();
+
+    /**
+     * Constructs an angle from a value in radians.
+     *
+     * @param rad the angle in radians.
+     *
+     * @return the constructed angle
+     */
+    static constexpr Angle ofRadians(double rad);
```
Comment (jonathanlew) [inline src/software/new_geom/angle.h]: I'd prefer fromRadians

```suggestion
    static constexpr Angle fromRadians(double rad);
```

File: src/software/new_geom/angle.h [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+
+    /**
+     * Constructs an angle from a value in degrees.
+     *
+     * @param deg the angle in degrees
+     *
+     * @return the constructed angle
+     */
+    static constexpr Angle ofDegrees(double deg);
```
Comment (jonathanlew) [inline src/software/new_geom/angle.h]: same here

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looks good! Have you evaluate how much will need to change to integrate this?

File: src/software/new_geom/point.h [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+     * @param the Point to duplicate
+     */
+    constexpr Point(const Point &p);
+
+    /**
+     * Creates a new Point from a Vector
+     *
+     * @param the Vector to create a Point from
+     */
+    constexpr Point(const Vector &v);
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/new_geom/point.h]: I thought we were not going to allow `Point + Point`? We should definitely declare this constructor `explicit` though to force the user to call it (https://stackoverflow.com/questions/121162/what-does-the-explicit-keyword-mean)

File: src/software/new_geom/angle.h (hunk 1 above)
Comment (garethellis0) [reply src/software/new_geom/angle.h]: Agreed, this has annoyed me for a while.

File: src/software/new_geom/angle.h [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+
+    /**
+     * Computes the arc sine of a value.
```
Comment (garethellis0) [inline src/software/new_geom/angle.h]:
``` suggestion
     * Computes the arc sin of a value.
```

File: src/software/new_geom/angle.h [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+     *
+     * The angle is rotated by a multiple of 2π until it lies within the target
+     * interval.
+     *
+     * @return the clamped angle.
+     */
+    constexpr Angle angleMod() const;
+
+   private:
+    double rads;
```
Comment (garethellis0) [inline src/software/new_geom/angle.h:185]: We should add a jdoc comment here.

File: src/software/new_geom/angle.h [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+inline std::ostream &operator<<(std::ostream &os, const Angle &a)
+{
+    os << a.toRadians() << "R";
+    return os;
+}
+
+// We also use variables of type 'Angle' to represent angular velocities, since they
+// are essentially represented the same. This typedef allows us to refer to Angles as
+// AngularVelocities, which makes the interfaces more intuitive.
+typedef Angle AngularVelocity;
```
Comment (garethellis0) [inline src/software/new_geom/angle.h]: can we put this in a `angular_velocity.h` file since it's not really part of the `Angle` class?

File: src/software/new_geom/angle_test.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+TEST(AngleTest, asin)
+{
+    EXPECT_DOUBLE_EQ(0, Angle::asin(0).toRadians());
+    EXPECT_DOUBLE_EQ(M_PI / 2, Angle::asin(1).toRadians());
+    EXPECT_DOUBLE_EQ(-M_PI / 2, Angle::asin(-1).toRadians());
+    EXPECT_DOUBLE_EQ(30, Angle::asin(0.5).toDegrees());
+    EXPECT_NEAR(-2.865983983, Angle::asin(-0.05).toDegrees(), 1e-7);
+    EXPECT_NEAR(57.14011962, Angle::asin(0.84).toDegrees(), 1e-7);
+}
```
Comment (garethellis0) [inline src/software/new_geom/angle_test.cpp:222]: looks like we're missing tests for several functions here, (`<=`, `>`, `tan`, etc.)

File: src/software/new_geom/geom_constants.h [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    // Due to internal representation of doubles being slightly less accurate/consistent
+    // with some numbers and operations, we consider points that are very close together
+    // to be equal (since they likely are, just possibly slightly misrepresented by the
+    // system/compiler). We use this EPSILON as a threshold for comparison. 1e-15 was
+    // chosen as a value because doubles have about 16 consistent significant figures.
+    // Comparing numbers with 15 significant figures gives us a
+    // small buffer while remaining as accurate as possible.
+    // http://www.cplusplus.com/forum/beginner/95128/
+    static constexpr double EPSILON = 1e-15;
+    static constexpr double NEAR    = 1e-9;
```
Comment (garethellis0) [inline src/software/new_geom/geom_constants.h]: I'm OK with epsilon since it has a meaning derived from an actual limitation of the computer system, but `NEAR` seems way too vague and subjective to me.

File: src/software/new_geom/vector.h [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+}
+
+inline double Vector::len() const
+{
+    return std::hypot(_x, _y);
+}
+
+inline Vector Vector::norm() const
+{
+    return len() < 1.0e-9 ? Vector() : Vector(_x / len(), _y / len());
```
Comment (garethellis0) [inline src/software/new_geom/vector.h]: Why are we rounding down like this?

File: src/software/new_geom/vector.h [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+}
+
+inline Vector Vector::norm(double length) const
+{
+    return len() < 1.0e-9 ? Vector() : Vector(_x * length / len(), _y * length / len());
```
Comment (garethellis0) [inline src/software/new_geom/vector.h]: Why are we rouding down like this?

File: src/software/new_geom/vector_test.cpp [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+{
+    Vector u = Vector();
+    Vector v = Vector(0, 0);
+    EXPECT_EQ(u, v);
+
+    v.setX(3);
+
+    EXPECT_FALSE(u == v);
+    EXPECT_TRUE(u != v);
+}
```
Comment (garethellis0) [inline src/software/new_geom/vector_test.cpp:237]: Are we testing everything here, including all the operators?

File: src/software/new_geom/point_test.cpp [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+{
+    Point p = Point();
+    Point q = Point(0, 0);
+    EXPECT_EQ(p, q);
+
+    q.setX(3);
+
+    EXPECT_FALSE(p == q);
+    EXPECT_TRUE(p != q);
+}
```
Comment (garethellis0) [inline src/software/new_geom/point_test.cpp:120]: Are we testing everything here, including all the operators?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, no major comments, except for perhaps testing.

File: src/software/new_geom/point.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/new_geom/point.h]: Adding origin to vector would be point + vector = point eh?

File: src/software/new_geom/angle.h (hunk 4 above)
Comment (jonathanlew) [reply src/software/new_geom/angle.h]: Sine is correct

Comment (jonathanlew) [reply src/software/new_geom/angle.h]: https://en.m.wikipedia.org/wiki/Sine

File: src/software/new_geom/point_test.cpp (hunk 12 above)
Comment (scveloso) [reply src/software/new_geom/point_test.cpp:120]: Yep every constructor/function/operator should be tested

File: src/software/new_geom/vector_test.cpp (hunk 11 above)
Comment (scveloso) [reply src/software/new_geom/vector_test.cpp:237]: Yep every constructor/function/operator should be tested

File: src/software/new_geom/geom_constants.h (hunk 8 above)
Comment (scveloso) [reply src/software/new_geom/geom_constants.h]: Should we leave NEAR as `1e-9` in it's usages?

File: src/software/new_geom/angle_test.cpp (hunk 7 above)
Comment (scveloso) [reply src/software/new_geom/angle_test.cpp:222]: ill fill these out in the next update

File: src/software/new_geom/point.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/point.h]: > I thought we were not going to allow Point + Point?

Point + Point is not allowed, I didn't implement it. What is implemented is Point + Vector, which should be allowed IMO

**Review by scveloso — COMMENTED** [review]
> Looks good! Have you evaluate how much will need to change to integrate this?

I looked through usages of old Point/Vector and they're used interchangeably but it seems to me there's always a more appropriate thing to use whether it is Point or Vector. It will be a fair bit of work to integrate this but that will be on future me :)

**Review by jsn0716 — COMMENTED** [review]
Looks good to me, except the build seems to be failing?

File: src/software/new_geom/vector.h (hunk 9 above)
Comment (MathewMacDougall) [reply src/software/new_geom/vector.h]: I suspect it's from the old implementation. I can't think of any reason other than numerical stability of some kind? Otherwise you could end up with "infinitely small" vectors? which if you're doing operation on the length could cause values to blow up?

File: src/software/new_geom/point.h [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+     * @return the distance
+     */
+    double distanceFromOrigin() const;
+
+    /**
+     * Returns the distance between this Point and the given point p
+     *
+     * @return the distance
+     */
+    double distanceFromPoint(const Point &p) const;
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h:93]: Is this redundant with the `dist(Point, Point)` function in `geom/util.h`?

File: src/software/new_geom/point.h [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+     *
+     * @param length the desired length of the resultant vector
+     *
+     * @return a vector in the same direction as this Point and with the given length,
+     * or a zero-vector if this Point is zero
+     */
+    Vector norm(double length) const;
+
+    /**
+     * Returns a new Point that is this Point rotated counterclockwise by an angle
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h:103]: Lets specify the rotation happens about the origin

File: src/software/new_geom/point.h [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+     *              | Y
+     *              |
+     *              |
+     *      +---------------+
+     *       -X     |      X
+     *              |
+     *              | -Y
+     *              +
+     */
+    Angle orientation() const;
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h]: Should we be able to get orientation from a Point? I'd say this can probably be removed?

File: src/software/new_geom/point.h [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+
+    /**
+     * Checks whether this Point contains NaN in either coordinate
+     *
+     * @return true if either coordinate is NaN, and false otherwise
+     */
+    constexpr bool isnan() const;
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h]: Thoughts on this function? There's currently no usages of it, so I think we can remove it. Not to say it's not useful, but we already aren't using it anywhere and generally should be avoiding cases where things become nan. Unit tests should catch any major issues.

File: src/software/new_geom/point.h [hunk 17]
```diff
@@ -0,0 +1,451 @@
...
+
+/**
+ * Adds a vector to a point
+ *
+ * @param p the Point
+ * @param v the Vector
+ *
+ * @return the Point sum of the given Point and Vector
+ */
+constexpr Point operator+(const Point &p, const Vector &v);
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h]: Can we add `__attribute__((warn_unused_result));`?

File: src/software/new_geom/point.h [hunk 18]
```diff
@@ -0,0 +1,451 @@
...
+    explicit constexpr Point();
+
+    /**
+     * Creates a unit-magnitude Point from an angle.
+     *
+     * @param angle the angle
+     *
+     * @return Point the Point
+     */
+    static Point createFromAngle(const Angle &angle);
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h]: Does it make sense for Points to be created from Angle? I would think this is more appropriate for Vector.

File: src/software/new_geom/point.h [hunk 19]
```diff
@@ -0,0 +1,451 @@
...
+     * @param y the <var>y</var> value of the Point
+     */
+    constexpr Point(double x, double y);
+
+    /**
+     * Creates a new Point that is a copy of the given Point
+     *
+     * @param the Point to duplicate
+     */
+    constexpr Point(const Point &p);
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h]: Can we add the `explicit` keyword here too?

File: src/software/new_geom/point.h [hunk 20]
```diff
@@ -0,0 +1,451 @@
...
+     * @return this Point
+     */
+    Point &operator=(const Point &other);
+
+   private:
+    /**
+     * The X coordinate of the Point. The variable name starts with an underscore to
+     * prevent name conflicts with its accessor function.
+     */
+    double _x;
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h]: super duper nit: Generally if we have to add an underscore to a member variable so that it doesn't match member functions, we append it (ie. `x_` rather than `_x`).

(same for `_y`)

File: src/software/new_geom/vector.h [hunk 21]
```diff
@@ -0,0 +1,521 @@
...
+     * @param y the new y-coordinate magnitude
+     */
+    void setY(double y);
+
+    /**
+     * Returns the square of the length of the Vector
+     *
+     * @return the square of the length of the Vector
+     */
+    constexpr double lensq() const;
```
Comment (MathewMacDougall) [inline src/software/new_geom/vector.h]: Do we want to take this opportunity to un-shorten these function names so things are more readable? IIRC we thought of this a while ago but it wasn't needed at the time. @garethellis0 ?

File: src/software/new_geom/vector.h [hunk 22]
```diff
@@ -0,0 +1,521 @@
...
+     *              +
+     */
+    Angle orientation() const;
+
+    /**
+     * Checks whether this Vector contains NaN in either of its x or y magnitude
+     *
+     * @return true if either x or y magnitude is NaN, and false otherwise
+     */
+    constexpr bool isnan() const;
```
Comment (MathewMacDougall) [inline src/software/new_geom/vector.h]: Same as Point, this isn't used so likely can be removed.

File: src/software/new_geom/vector.h [hunk 23]
```diff
@@ -0,0 +1,521 @@
...
+     * @return this Vector
+     */
+    Vector &operator=(const Vector &other);
+
+   private:
+    /**
+     * The magnitude in the X coordinate of the Vector. The variable name starts with an
+     * underscore to prevent name conflicts with its accessor function.
+     */
+    double _x;
```
Comment (MathewMacDougall) [inline src/software/new_geom/vector.h]: Same here for member variable naming.

File: src/software/new_geom/vector.h [hunk 24]
```diff
@@ -0,0 +1,521 @@
...
+
+/**
+ * Adds two vectors
+ *
+ * @param p the first Vector
+ * @param q the second Vector
+ *
+ * @return the vector-sum of the two vectors
+ */
+constexpr Vector operator+(const Vector &p, const Vector &q);
```
Comment (MathewMacDougall) [inline src/software/new_geom/vector.h]: Can we add `__attribute__((warn_unused_result));`?

File: src/software/new_geom/vector.h [hunk 25]
```diff
@@ -0,0 +1,521 @@
...
+
+/**
+ * Multiplies a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scaling factor
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator*(const Vector &p, double s);
```
Comment (MathewMacDougall) [inline src/software/new_geom/vector.h]: Can we add `__attribute__((warn_unused_result));`

File: src/software/new_geom/vector.h [hunk 26]
```diff
@@ -0,0 +1,521 @@
...
+
+/**
+ * Divides a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scalar to divide by
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator/(const Vector &p, double s);
```
Comment (MathewMacDougall) [inline src/software/new_geom/vector.h]: Can we add `__attribute__((warn_unused_result));`?

File: src/software/new_geom/geom_constants.h (hunk 8 above)
Comment (MathewMacDougall) [reply src/software/new_geom/geom_constants.h]: Seems like it's only used to check if a Point is near another point (ie. they're very close but not equal). I don't know if we use `isClose` without any parameters, so maybe this variable and the `isClose()` function can be removed?

File: src/software/new_geom/point.h [hunk 27]
```diff
@@ -0,0 +1,451 @@
...
+     */
+    Vector toVector() const;
+
+    /**
+     * Returns the unit vector in direction of this point
+     *
+     * @return A unit vector in the direction of this Point, or a
+     * zero-vector if this Point is zero
+     */
+    Vector norm() const;
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h]: Does it make sense to normalize a Point? Are we intending to return a `Vector` here?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looking good? Most comments are about what functions should be available for each class.

File: src/software/new_geom/geom_constants.h (hunk 8 above)
Comment (garethellis0) [reply src/software/new_geom/geom_constants.h]: Yeah I was gonna say that we should probably reconsider any functions that use `NEAR`. I'd say remove `isClose()` and see if we can do without.

**Comment by scveloso** [comment]
**Latest update to PR addresses comments:**
- `_x` and `_y` to `x_` and `y_`
- Fully test `angle.h`
- Remove `isnan()` from `point.h` and `vector.h`
- Remove `NEAR` from `geom_constants.h` and `isClose()` from `point.h` since no usages and too subjective
- Add `__attribute__((warn_unused_result));` to `point.h` and `vector.h` operators that should have it
- Add `explicit` to `Point(Point p)` and `Point(Vector v)`
- Moved `AngularVelocity` to its own header
- Some other minor stuff

~~**Outstanding comments:**~~
~~- Should we have `orientation()` in `Point`?~~
~~- Should we have `norm()` in `Point`?~~
~~- Should `Point` be created from an `Angle`?~~

~~Both of these I imagined a `Point` on a unit circle and it seemed appropriate to have those methods. But at the same time, these seem more `Vector`-y than `Point`-y. I'm now leaning towards removing them from `Point`.~~

**Comment by garethellis0** [comment]
I don't think we should have `orientation()` or `norm()`, nor should we be able to create a `Point` from an `Angle`.

**Comment by garethellis0** [comment]
@scveloso - please just re-request review whenever you want another (little arrow circle by each persons name under "Reviewers")

File: src/software/new_geom/vector.h (hunk 9 above)
Comment (garethellis0) [reply src/software/new_geom/vector.h]: I mean, a zero length vector would have the same effect as an "infinitely small" one. If we were really in that situation, I'd say it's the job of the user to compare the vector norm to `EPSILON` (but definitely not `NEAR`)

File: src/software/new_geom/angular_velocity.h [hunk 28]
```diff
@@ -0,0 +1,6 @@
+#include "software/new_geom/angle.h"
+
+// We also use variables of type 'Angle' to represent angular velocities, since they
+// are essentially represented the same. This typedef allows us to refer to Angles as
+// AngularVelocities, which makes the interfaces more intuitive.
+typedef Angle AngularVelocity;
```
Comment (garethellis0) [inline src/software/new_geom/angular_velocity.h]: missing newline at end of file

File: src/software/new_geom/point.h [hunk 29]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * Checks whether this Point is close to another Point, where “close”
+     * is defined as 1.0e-9
+     *
+     * @param other the other point to check against
+     *
+     * @return true if the other point is within a distance of 1.0e-9, not inclusive
+     */
+    bool isClose(const Point &other) const;
```
Comment (MathewMacDougall) [inline src/software/new_geom/point.h]: Did you mean to remove this?

File: src/software/new_geom/vector.h [hunk 30]
```diff
@@ -0,0 +1,510 @@
...
+     *
+     * @param length the desired length of the resultant vector
+     *
+     * @return a vector in the same direction as this Vector and with the given length,
+     * or a zero-vector if this Vector is a zero-vector
+     */
+    Vector norm(double length) const;
+
+    /**
+     * Returns the vector perpendicular to this Vector
```
Comment (MathewMacDougall) [inline src/software/new_geom/vector.h]: Do we want to formalize any convention for which perpendicular vector is returned here? I happen to know it's rotated positive 90 degrees from the original vector (so pointing perpendicular left), so should we formalize that in the comment or do people have other opinions on how perpendicular vectors should work?

@garethellis0 @jonathanlew @jsn0716

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Awesome! Just couple small convention things I missed

File: src/software/new_geom/point.h (hunk 14 above)
Comment (MathewMacDougall) [reply src/software/new_geom/point.h:103]: bumping this comment

File: src/software/new_geom/point.h (hunk 13 above)
Comment (MathewMacDougall) [reply src/software/new_geom/point.h:93]: I decided I don't have a strong opinion about this, unless @garethellis0 or @jonathanlew do?

**Review by jonathanlew — DISMISSED** [review]
Nothing else beyond what others have mentioned. Looking good!

**Comment by MathewMacDougall** [comment]
Awesome, LGTM. Just waiting for @garethellis0 opinion on my last comment about shortened names.

File: src/software/new_geom/vector.h (hunk 30 above)
Comment (JasonCPEN221) [reply src/software/new_geom/vector.h]: Has this been resolved already? I think it should either be specified in the comment to be a positive 90 degrees rotation or take a bool parameter to rotate one way or the other.

Comment (jsn0716) [reply src/software/new_geom/vector.h]: Has this been resolved? I think it should either be specified in the comment that it's a positive rotation or there should be a bool parameter to decide which direction

Comment (MathewMacDougall) [reply src/software/new_geom/vector.h]: Yes it's resolved. GitHub doesn't show the changes in the conversation comment, it just marks it as "outdated" at the top. If you look at the latest set of changes it should show the new comment.

File: src/software/new_geom/vector.h (hunk 21 above)
Comment (garethellis0) [reply src/software/new_geom/vector.h]: yes, absolutely. This should be `lengthSquared`

Comment (garethellis0) [reply src/software/new_geom/vector.h]: Same sort of thing goes if there are other similar names.

**Comment by garethellis0** [comment]
@scveloso - would you mind leaving a quick comment reply on each comment you resolved? Even if it's as short as "changed" or "not changed bcause ..." it really helps me when I'm going back over PR's.

File: src/software/new_geom/angular_velocity.h (hunk 28 above)
Comment (scveloso) [reply src/software/new_geom/angular_velocity.h]: fixed in latest update

File: src/software/new_geom/geom_constants.h (hunk 8 above)
Comment (scveloso) [reply src/software/new_geom/geom_constants.h]: Removed `isClose()` and `NEAR` in the latest update

File: src/software/new_geom/angle_test.cpp (hunk 7 above)
Comment (scveloso) [reply src/software/new_geom/angle_test.cpp:222]: fixed in latest update

File: src/software/new_geom/point.h (hunk 29 above)
Comment (scveloso) [reply src/software/new_geom/point.h]: Yep sorry, removed in latest update

File: src/software/new_geom/angle.h [hunk 31]
```diff
@@ -0,0 +1,586 @@
...
+};
+
+/**
+ * Negates an angle.
+ *
+ * @param angle the angle to negate.
+ *
+ * @return the negated angle
+ */
+constexpr Angle operator-(const Angle &angle);
```
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/angle.h]: Addressed in latest update

**Comment by scveloso** [comment]
> would you mind leaving a quick comment reply on each comment you resolved? Even if it's as short as "changed" or "not changed bcause ..." it really helps me when I'm going back over PR's.

@garethellis0 sorry was a little busy when updating the PR, I updated the recent ones and will remember to do so next time

**Comment by garethellis0** [comment]
Quick request. If you have a sec before tomorrow, could you do some performance testing for the `Vector` class? Namely, I was talking to Mat, and we're fairly sure the compiler will just inline all the functions, so we shouldn't need to declare them inline in the header (I believe it's just supposed to be a performance optimization). The test I'm imagining is running a bunch of functions from the vector class some really large (think 10^8 sorta scale) number of times, and seeing if there's any significant time difference.

If there's not a significant difference, we should put all the implementations of the functions into a `.cpp` file, which would make the header much less extreme (it's 500+ lines :grimacing: ).

This has actually been a subject of much debate over the last few years, would be good to get some data one way or the other.

File: src/software/new_geom/vector.h [hunk 32]
```diff
@@ -0,0 +1,511 @@
...
+}
+
+inline double Vector::len() const
+{
+    return std::hypot(x_, y_);
+}
+
+inline Vector Vector::norm() const
+{
+    return len() < 1.0e-9 ? Vector() : Vector(x_ / len(), y_ / len());
```
Comment (garethellis0) [inline src/software/new_geom/vector.h]: Is there a specific reason we're using `1.0e-9` here? I would imagine we'd want to compare to `EPSILON` instead (ie. check that the number is basically 0 in terms of what the computer is capable of representing).

File: src/software/new_geom/vector.h (hunk 10 above)
Comment (garethellis0) [reply src/software/new_geom/vector.h]: Actually the rounding somewhat makes sense, but is there a specific reason we're using `1.0e-9` here? I would imagine we'd want to compare to `EPSILON` instead (ie. check that the number is basically 0 in terms of what the computer is capable of representing).

File: src/software/new_geom/vector.h [hunk 33]
```diff
@@ -0,0 +1,511 @@
...
+}
+
+inline Vector Vector::norm(double length) const
+{
+    return len() < 1.0e-9 ? Vector() : Vector(x_ * length / len(), y_ * length / len());
```
Comment (garethellis0) [inline src/software/new_geom/vector.h]: Can simplify to `norm() * length` I think?

File: src/software/new_geom/vector_test.cpp [hunk 34]
```diff
@@ -0,0 +1,228 @@
...
+
+TEST(VectorLogicTests, norm_vector_from_vector_with_length_test)
+{
+    Vector u = Vector(3, 4);
+    Vector v = u.norm(2);
+    EXPECT_EQ(1.2, v.x());
+    EXPECT_EQ(1.6, v.y());
+}
+
```
Comment (garethellis0) [inline src/software/new_geom/vector_test.cpp:81]: Could we add a test for taking the `norm()` of a vector that is just off of zero (but less then `EPSILON`)?

**Comment by garethellis0** [comment]
@scveloso - looking good! Two quick comments on the `norm()` function and one suggested test for it, other then that, this looks good to go from my end.

File: src/software/new_geom/vector.h (hunk 10 above)
Comment (MathewMacDougall) [reply src/software/new_geom/vector.h]: `1e-9` was the value of `NEAR`.... :roll_eyes:

**Comment by MathewMacDougall** [comment]
For the speed testing, make sure to test compiling in both debug and release mode.

**Comment by jonathanlew** [comment]
> Quick request. If you have a sec before tomorrow, could you do some performance testing for the `Vector` class? Namely, I was talking to Mat, and we're fairly sure the compiler will just inline all the functions, so we shouldn't need to declare them inline in the header (I believe it's just supposed to be a performance optimization). The test I'm imagining is running a bunch of functions from the vector class some really large (think 10^8 sorta scale) number of times, and seeing if there's any significant time difference.
> 
> If there's not a significant difference, we should put all the implementations of the functions into a `.cpp` file, which would make the header much less extreme (it's 500+ lines  ).
> 
> This has actually been a subject of much debate over the last few years, would be good to get some data one way or the other.

Why can't we inline from a cpp file?

**Comment by garethellis0** [comment]
@jonathanlew - don't think you can: https://stackoverflow.com/questions/3992980/c-inline-member-function-in-cpp-file?

File: src/software/new_geom/vector.h (hunk 10 above)
Comment (garethellis0) [reply src/software/new_geom/vector.h]: Haha, I think this is a good opportunity to question some of the decisions that were made when designing this originally. If we can't come up with a good reason for this value, I'd be inclined to use `EPSILON` (because that indicates 0 to within machine precision).

File: src/software/new_geom/vector_test.cpp (hunk 34 above)
Comment (scveloso) [reply src/software/new_geom/vector_test.cpp:81]: updated

File: src/software/new_geom/vector.h (hunk 33 above)
Comment (scveloso) [reply src/software/new_geom/vector.h]: fixed

**Comment by scveloso** [comment]
> Quick request. If you have a sec before tomorrow, could you do some performance testing for the `Vector` class? Namely, I was talking to Mat, and we're fairly sure the compiler will just inline all the functions, so we shouldn't need to declare them inline in the header (I believe it's just supposed to be a performance optimization). The test I'm imagining is running a bunch of functions from the vector class some really large (think 10^8 sorta scale) number of times, and seeing if there's any significant time difference.
> 
> If there's not a significant difference, we should put all the implementations of the functions into a `.cpp` file, which would make the header much less extreme (it's 500+ lines  ).
> 
> This has actually been a subject of much debate over the last few years, would be good to get some data one way or the other.

@garethellis0 I ran all the vector functions 10^8 times, twice, one using inline keyword and another without the inline keyword. 

With inline: 161.567s
Without inline: 159.314s

Was this what you meant? Or did you have another test in mind?

**Comment by scveloso** [comment]
@garethellis0 one more thing to consider is ```constexpr``` implies ```inline``` so moving some functions to a .cpp file from a .h means we lose the ```constexpr```

We could just move non-constexpr functions but as it is, that wouldn't reduce the size of the .h file by much

**Review by jsn0716 — CHANGES_REQUESTED** [review]
One nit and there seems to be an unresolved review for `vector.h` regarding function names. Otherwise LGTM

**Comment by garethellis0** [comment]
@scveloso - I think we can deal with losing the `constexpr` qualifiers. Maybe do one final speed test with everything in the `.cpp`, but I think we should be good to move it all. Thanks for taking the time to do those tests!

**Comment by scveloso** [comment]
@garethellis0 did another test running all vector functions 10^8 times for both header+inline keyword (old geom way of doing it) and separating .h and .cpp file:
- t_header_inline = 157.883s
- t_cpp_file = 162.526s

Ill update the PR soon with the .h and .cpp separated

File: src/software/new_geom/geom_constants.h [hunk 35]
```diff
@@ -0,0 +1,12 @@
+namespace GeomConstants
+{
+    // Due to internal representation of doubles being slightly less accurate/consistent
+    // with some numbers and operations, we consider points that are very close together
```
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/geom_constants.h]: fixed

**Review by MathewMacDougall — APPROVED** [review]
Nice, LGTM :+1: Thanks for persevering through this

**Review by jsn0716 — APPROVED** [review]
LGTM 👍

**Review by jonathanlew — CHANGES_REQUESTED** [review]
If performance isn't an issue, we should move angle definitions to angle.cpp, too. @MathewMacDougall @garethellis0 thoughts?

**Comment by MathewMacDougall** [comment]
Yes I'd like to in that case, but that can be a separate PR.

**Comment by jonathanlew** [comment]
> Yes I'd like to in that case, but that can be a separate PR.

kk

**Review by jonathanlew — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]
Great work with this! LGTM :+1: :tada:
